### PR TITLE
Use v2 registry API

### PR DESCRIPTION
**Pull request by fgimenez**
Uses registry v2 API for getting existing image tags on the target registry. Includes code for requesting a token according to https://docs.docker.com/registry/spec/auth/token/. Tested with Quay and aliyun.

### CODE REVIEW

File: registry.go [hunk 1]
```diff
@@ -75,12 +85,18 @@ func (r *Registry) Login() error {
 }
 
 func (r *Registry) CheckImageTagExists(image, tag string) (bool, error) {
-	url := fmt.Sprintf("https://%s/api/v1/repository/%s/tag/%s/images", r.host, ImageName(r.organisation, image), tag)
+	url := fmt.Sprintf("https://%s/v2/%s/tags/list", r.host, ImageName(r.organisation, image))
```
Comment (rossf7) [inline registry.go]: <3

File: registry.go [hunk 2]
```diff
@@ -108,3 +134,80 @@ func (r *Registry) Retag(image, sha, tag string) (string, error) {
 	}
 	return retaggedNameWithTag, nil
 }
+
+func (r *Registry) getToken(req *http.Request) (string, error) {
+	const authorizationHeaderKey = "Www-Authenticate"
```
Comment (rossf7) [inline registry.go]: Does the casing here matter? I did a quick google and saw some examples also using `WWW-Authenticate`.

File: registry_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+func Test_GetAuthURL(t *testing.T) {
+	tcs := []struct {
+		challenge     string
+		expectedURL   string
+		expectedError bool
+		description   string
+	}{
+		{
+			description: "base case",
```
Comment (rossf7) [inline registry_test.go]: Nitpick but could we use the "case 0:" style we have in `fmt`? 

I find it quicker to track down test failures.

File: registry.go (hunk 2 above)
Comment (fgimenez) [reply registry.go]: Very good point, on aliyun we have `Www-Authenticate` but even in the RFC they talk about `WWW-Authenticate` https://tools.ietf.org/html/rfc6750#section-3. I've changed the code to iterate through the header keys and check against a lowercased version, PTAL.

File: registry_test.go (hunk 3 above)
Comment (fgimenez) [reply registry_test.go]: Sure, done.

File: registry.go (hunk 2 above)
Comment (rossf7) [reply registry.go]: Nice! Good that the comparison is now case insensitive. I can definitely imagine some other registry using a different case and that could be nasty to debug!

**Pull request merged**